### PR TITLE
RTCP-XR/Vq Reporting Support (PUBLISH)

### PR DESCRIPTION
**Pull request by lmangani**
This *DIRTY, EXPERIMENTAL PR (mostly intended for peer review)* implements support in Baresip for RTCP-XR/Vq (RFC6035) call quality reports as PUBLISH messages and implements a rudimentary RTCP MOS scoring mechanism to complement them, just like fancy Hardware devices. As per *rtpstats*, a dedicated config pair controls the feature activation *(rtcpxr_stats)* and SIP target for the reports *(rtcpxr_collector)*
```
rtp_stats             no
rtcpxr_stats          no
rtcpxr_collector      sip:rtcpxr@127.0.0.1:5088
```
NOTE: This code was developed for Baresip 0.4.18 and just re-synced with the latest GIT version, only tested with basic scenarios w/ the precious help of @alfredh but I unfortunately never had time to polish it, so I'm posting this PR here in case anyone is interested in giving this a shot and/or help correct/improve the code to allow a safe merge or branch - *thanks in advance to any brave soul!*

Functionality is contributed by and validated with the [HOMER](http://sipcapture.org) Project but should be generally compatible with any other RTCP-XR/Vq collector.

image

### CODE REVIEW

File: src/rtcpxr.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+			   str_len(msg), msg);
+
+	mem_deref(uri);
+
+	return err;
+}
+
+void rtcpxr_test(void){
+		re_fprintf(stderr, "RTCP-XR Test Function Call \n");
+}
```
Comment (alfredh) [inline src/rtcpxr.c]: this function look unused, perhaps it can be removed ?

File: src/audio.c [hunk 2]
```diff
@@ -1995,3 +1995,128 @@ int audio_print_rtpstat(struct re_printf *pf, const struct audio *a)
...
+  else if (*r_factor < 0) *r_factor = 0;
+  mos_val = 1 + (0.035) * (*r_factor) +
+	(0.000007) * (*r_factor) * ((*r_factor) - 60)
+	* (100 - (*r_factor));
+  if (mos_val > 5) mos_val = 5;
+/* info("[RTT: %.2f][Jitter: %.2f][# Packet Lost: %u][R-Factor: %.2f]"
+	"[MOS: %.2f] \n", rtt, jitter, num_packets_lost,
+	 *r_factor, mos_val); */
+  return(mos_val);
+}
```
Comment (alfredh) [inline src/audio.c]: there is a new function in baresip to calculate MOS:

```
double mos_calculate(double *r_factor, double rtt,
		     double jitter, uint32_t num_packets_lost);
```

can this one be used instead ?

File: src/account.c [hunk 3]
```diff
@@ -360,9 +360,10 @@ int account_alloc(struct account **accp, const char *sipaddr)
 		goto out;
 
 	pl_set_str(&pl, acc->buf);
+
 	err = sip_addr_decode(&acc->laddr, &pl);
 	if (err) {
-		warning("account: invalid SIP address: `%r'\n", &pl);
+		warning("account: error parsing SIP address: '%r'\n", &pl);
```
Comment (alfredh) [inline src/account.c]: applied in a separate commit.

File: src/audio.c (hunk 2 above)
Comment (lmangani) [reply src/audio.c]: It definitely should - would you have time to do this hookup?

File: src/rtcpxr.c [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+ * Send SIP instant PUBLISH RTCP-XR to a peer
+ *
+ * @param ua    User-Agent object
+ * @param peer  Peer SIP Address
+ * @param msg   Message to send
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int rtcpxr_send(struct ua *ua, const char *peer, const char *msg, ...)
+{
```
Comment (alfredh) [inline src/rtcpxr.c]: `msg` is a formatted string and you can't really just send it into `sip_req_send()`

a good example of how to format the string is `re_sdprintf`:

```
int re_sdprintf(char **strp, const char *fmt, ...)
{
	va_list ap;
	int err;

	va_start(ap, fmt);
	err = re_vsdprintf(strp, fmt, ap);
	va_end(ap);

	return err;
}
```

File: src/call.c [hunk 5]
```diff
@@ -732,13 +732,22 @@ int call_modify(struct call *call)
 int call_hangup(struct call *call, uint16_t scode, const char *reason)
 {
 	int err = 0;
+	char msg[512];
+
+	memset(msg,0,512);
```
Comment (alfredh) [inline src/call.c]: is this variable unused ?

**Comment by alfredh** [comment]
hi @lmangani many thanks for your PR,

couple of comments so far:

* we should add a selftest for this feature. The selftest mock can implement a
  dummy PUBLISH handler that is used as the rtcpxr collector

* could you please rebase the patch on top of master and also squash all
  commits into one ?

* could you please try selftest with this PR and see if it works ?
  (`make test`)


Alfred

**Comment by lmangani** [comment]
Hi @alfredh 

Thanks for taking time to review the existing code, much appreciated! As of the points, my main issue is time as indicated in the PR description - I'll sure try to find the time to do so. If anyone in the community is interested in this feature, and willing to step in and spare a hand with the test part, please feel free to do so!

File: include/baresip.h [hunk 6]
```diff
@@ -230,6 +230,8 @@ struct config_avt {
 	bool rtcp_mux;          /**< RTP/RTCP multiplexing          */
 	struct range jbuf_del;  /**< Delay, number of frames        */
 	bool rtp_stats;         /**< Enable RTP statistics          */
+	bool rtcpxr_stats;      /**< Enable RTCP-XR statistics      */
+	char rtcpxr_collector[128]; /**< RTCP-XR Publish collector  */
```
Comment (alfredh) [inline include/baresip.h]: should the rtcpxr_collector be specified per UA instead ?

File: include/baresip.h [hunk 7]
```diff
@@ -354,6 +356,18 @@ int  message_send(struct ua *ua, const char *peer, const char *msg,
 		  sip_resp_h *resph, void *arg);
 
 
+/*
+ * PUBLISH RTCP-XR
+ */
+
+typedef void (rtcpxr_recv_h)(const struct pl *peer, const struct pl *ctype,
+			      struct mbuf *body, void *arg);
+int  rtcpxr_init(rtcpxr_recv_h *recvh, void *arg);
```
Comment (alfredh) [inline include/baresip.h]: rtcpxr_init is not called from anywhere.

do you also intend for baresip to *receive* PUBLISH events ?

File: include/baresip.h [hunk 8]
```diff
@@ -354,6 +356,18 @@ int  message_send(struct ua *ua, const char *peer, const char *msg,
...
+void rtcpxr_close(void);
+int  rtcpxr_send(struct ua *ua, const char *peer, const char *msg, ...);
+void  rtcpxr_test(void);
```
Comment (alfredh) [inline include/baresip.h]: rtcpxr_test is unused prototype.

File: src/audio.c [hunk 9]
```diff
@@ -1995,3 +1995,128 @@ int audio_print_rtpstat(struct re_printf *pf, const struct audio *a)
...
+
+
+/*
+ *
+ * RTCP-XR PUBLISH
+ * Reference:  RFC 6035
+ *
+ */
+char * audio_print_rtcpxr(char * msg, int msg_len,
+       const struct audio *a, const char * cid)
```
Comment (alfredh) [inline src/audio.c]: I think it is better to return `int` here for the error code.

perhaps we can use a `%H` compatible printer instead ?

File: src/call.c [hunk 10]
```diff
@@ -1817,6 +1829,18 @@ void call_set_xrtpstat(struct call *call)
 				  audio_print_rtpstat, call->audio);
 }
 
+void call_send_rtcpxr(struct call *call)
```
Comment (alfredh) [inline src/call.c]: check input argument, return EINVAL if NULL.

File: src/call.c [hunk 11]
```diff
@@ -1817,6 +1829,18 @@ void call_set_xrtpstat(struct call *call)
...
+{
+	char msg[512];
+	memset(msg,0,512);
+
+	/* info("call: Generating RTCP-XR Statistics for '%s'\n", */
+	/*     sip_dialog_callid(sipsess_dialog(call->sess))); */
+
+	rtcpxr_send(call->ua,call->config_avt.rtcpxr_collector,
+		audio_print_rtcpxr(msg,sizeof(msg),call->audio,
+		sip_dialog_callid(sipsess_dialog(call->sess)) ) );
```
Comment (alfredh) [inline src/call.c]: check return value

File: src/config.c [hunk 12]
```diff
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 2010 Creytiv.com
  */
+
```
Comment (alfredh) [inline src/config.c]: unrelated change

File: src/config.c [hunk 13]
```diff
@@ -283,7 +286,12 @@ int config_parse_conf(struct config *cfg, const struct conf *conf)
 	(void)conf_get_bool(conf, "rtcp_mux", &cfg->avt.rtcp_mux);
 	(void)conf_get_range(conf, "jitter_buffer_delay",
 			     &cfg->avt.jbuf_del);
+
+	/* QoS Settings */
 	(void)conf_get_bool(conf, "rtp_stats", &cfg->avt.rtp_stats);
+	(void)conf_get_bool(conf, "rtcpxr_stats", &cfg->avt.rtcpxr_stats);
+	(void)conf_get_str(conf, "rtcpxr_collector",
+	      cfg->avt.rtcpxr_collector, sizeof(cfg->avt.rtcpxr_collector));
```
Comment (alfredh) [inline src/config.c]: does the rtcpxr_stats config require rtcpxr_collector to be set?

the rtcpxr_collector string should be parsed and validated here.

File: src/rtcpxr.c [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+ * Copyright (C) 2010 Creytiv.com
+ */
+#include <re.h>
+#include <baresip.h>
+#include "core.h"
+
+
+static struct sip_lsnr *lsnr;
+static rtcpxr_recv_h *recvh;
+static void *recvarg;
```
Comment (alfredh) [inline src/rtcpxr.c]: We are now working to make baresip 100% re-entrant

this adds static/global data and should be avoided. it is better to put them in a struct
that is allocated.

**Comment by alfredh** [comment]
hi @lmangani 

some more comments inline and below:

- could this feature be implemented as an app_module instead ?
  if possible it would make the whole design much more flexible and elegant.

- checking return values must be done in all places.

- the rtcpxr_collector config should be parsed and checking during init

- when enabling rtcpxr_stats with an empy collector I get an error:
  call: rtcpxr_send error (Bad message)

- rtcpxr_init() is not called. is it needed ?

- is the RTCPXR valid for audio only, or also for video ?

- the PUBLISH message sent has no content:

U 127.0.0.1:55235 -> 127.0.0.1:55235
PUBLISH sip:b@127.0.0.1:55235 SIP/2.0.
Via: SIP/2.0/UDP 127.0.0.1:55235;branch=z9hG4bK3f8192e50998bfe2;rport.
Max-Forwards: 70.
To: <sip:b@127.0.0.1:55235>.
From: <sip:a@127.0.0.1>;tag=182c9430b1219f67.
Call-ID: 2154aeb18a060957.
CSeq: 17537 PUBLISH.
User-Agent: test.
Content-Type: application/vq-rtcpxr.
Content-Length: 0.

- should the "rtcpxr_collector" config be specified per ua instead?

- all new code should be re-entrant, ref rtcpxr.c


Then a couple of more questions:

* Could you please rebase the patch against master?

* Could you please squash all the commits to one single commit ?

Alfred

**Comment by lmangani** [comment]
Thanks @alfredh I'll take those points into consideration and attempt a rebase sometime early next year if the time allows.
